File: prob2.java
// Time Complexity : O(U+S)
// Space Complexity : O(S+G)

import java.util.*;

public class Main {

    public static Map<String, List<String>> favoriteGenre(Map<String, List<String>> userSongs, Map<String, List<String>> songGenres) {
        Map<String, String> songToGenre = new HashMap<>();
        for (Map.Entry<String, List<String>> entry : songGenres.entrySet()) {
            String genre = entry.getKey();
            List<String> songs = entry.getValue();
            for (String song : songs) {
                songToGenre.put(song, genre);
            }
        }
        
        Map<String, List<String>> userFavoriteGenres = new HashMap<>();
        for (Map.Entry<String, List<String>> entry : userSongs.entrySet()) {
            String user = entry.getKey();
            List<String> songs = entry.getValue();
            Map<String, Integer> genreCount = new HashMap<>();
            int maxCount = 0;
            for (String song : songs) {
                String genre = songToGenre.get(song);
                genreCount.put(genre, genreCount.getOrDefault(genre, 0) + 1);
                maxCount = Math.max(maxCount, genreCount.get(genre));
            }
            List<String> favoriteGenres = new ArrayList<>();
            for (Map.Entry<String, Integer> genreEntry : genreCount.entrySet()) {
                if (genreEntry.getValue() == maxCount) {
                    favoriteGenres.add(genreEntry.getKey());
                }
            }
            userFavoriteGenres.put(user, favoriteGenres);
        }
        
        return userFavoriteGenres;
    }

    public static void main(String[] args) {
        HashMap<String, List<String>> userSongs = new HashMap<>();

        userSongs.put("David", Arrays.asList(new String[]{"song1", "song2", "song3", "song4", "song8"}));
        userSongs.put("Emma", Arrays.asList(new String[]{"song5", "song6", "song7"}));

        HashMap<String, List<String>> songGenres = new HashMap<>();

        songGenres.put("Rock", Arrays.asList(new String[]{"song1", "song3"}));
        songGenres.put("Dubstep", Arrays.asList(new String[]{"song7"}));
        songGenres.put("Techno", Arrays.asList(new String[]{"song2", "song4"}));
        songGenres.put("Pop", Arrays.asList(new String[]{"song5", "song6"}));
        songGenres.put("Jazz", Arrays.asList(new String[]{"song8", "song9"}));

        Map<String, List<String>> res = favoriteGenre(userSongs, songGenres);

        System.out.println(res);
    }
}

File: prob2.py
# Time Complexity : O(mn)
# Space Complexity :O(n)
# Passed on Leetcode: yes

from collections import defaultdict

def favorite_genre(user_songs, song_genres):
    # Create a song to genre mapping for quick lookup
    song_to_genre = {}
    for genre, songs in song_genres.items():
        for song in songs:
            song_to_genre[song] = genre
    
    # Create a dictionary to store user's favorite genres
    user_favorite_genres = defaultdict(list)
    
    # Iterate through each user and their songs
    for user, songs in user_songs.items():
        genre_count = defaultdict(int)
        max_count = 0
        
        for song in songs:
            if song in song_to_genre:
                genre = song_to_genre[song]
                genre_count[genre] += 1
                max_count = max(max_count, genre_count[genre])
        
        for genre, count in genre_count.items():
            if count == max_count:
                user_favorite_genres[user].append(genre)
    
    return dict(user_favorite_genres)

# Test the function with your example data
user_songs = {
    "David": ["song1", "song2", "song3", "song4", "song8"],
    "Emma": ["song5", "song6", "song7"]
}

song_genres = {
    "Rock": ["song1", "song3"],
    "Dubstep": ["song7"],
    "Techno": ["song2", "song4"],
    "Pop": ["song5", "song6"],
    "Jazz": ["song8", "song9"]
}

result = favorite_genre(user_songs, song_genres)
print(result)
